### PR TITLE
Draw RelativePose2DStampedConstraint constraints

### PR DESCRIPTION
**Pull request by efernandez**
This includes:
* Dynamically generate display properties for the constraint sources
* Cache the constraint sources properties config so it's applied when
  the properties are later created
* Create Variable visual + property, as we do for the Constraint

Below you can see how the display looks like, its properties and visual objects, and what can be done to inspect a `fuse_msgs::SerializedGraph` message using this PR. Before only the variables were rendered.

simplescreenrecorder-2019-09-23_17 43 38

I've recently merged this same thing into our fork so we get some testing on it: https://github.com/clearpathrobotics/fuse/pull/2

Also note this requires `rviz` >= 1.13.4, but if that's an issue, I can make it work with a lower version. The reason is that I try to re-use code for the covariance property and visual, but it requires their code to be duplicated. With `rviz` >= 1.13.4, at least I don't have to duplicate the visual code, although I still have for the property one.

### CODE REVIEW

**Comment by efernandez** [comment]
@svwilliams Could you paste the test failure error here, please? I suspect it's related with the version of `rviz` being used, since 1.13.4 is the latest one and you probably have an older one in your test infrastructure.

**Comment by svwilliams** [comment]
```
[2019-09-26T15:40:19.753Z] Starting >>> fuse_viz

[2019-09-26T15:40:58.529Z] --- stderr: fuse_viz

[2019-09-26T15:40:58.529Z] In file included from /home/ec2-user/workspace/test_rosdistro_fuse_PR-107/package/fuse_viz/src/mapped_covariance_property.cpp:31:0:

[2019-09-26T15:40:58.529Z] /home/ec2-user/workspace/test_rosdistro_fuse_PR-107/package/fuse_viz/include/fuse_viz/mapped_covariance_visual.h: In constructor ‘rviz::MappedCovarianceVisual::MappedCovarianceVisual(Ogre::SceneManager*, Ogre::SceneNode*, bool, bool, float, float, float)’:

[2019-09-26T15:40:58.529Z] /home/ec2-user/workspace/test_rosdistro_fuse_PR-107/package/fuse_viz/include/fuse_viz/mapped_covariance_visual.h:74:115: error: ‘rviz::CovarianceVisual::CovarianceVisual(Ogre::SceneManager*, Ogre::SceneNode*, bool, bool, float, float, float)’ is private within this context

[2019-09-26T15:40:58.529Z]      : CovarianceVisual(scene_manager, parent_node, is_local_rotation, is_visible, pos_scale, ori_scale, ori_offset)

[2019-09-26T15:40:58.529Z]                                                                                                                    ^

[2019-09-26T15:40:58.529Z] In file included from /home/ec2-user/workspace/test_rosdistro_fuse_PR-107/package/fuse_viz/include/fuse_viz/mapped_covariance_visual.h:33:0,

[2019-09-26T15:40:58.529Z]                  from /home/ec2-user/workspace/test_rosdistro_fuse_PR-107/package/fuse_viz/src/mapped_covariance_property.cpp:31:

[2019-09-26T15:40:58.529Z] /opt/locusrobotics/hotdog/dev/ros1/include/rviz/default_plugin/covariance_visual.h:93:3: note: declared private here

[2019-09-26T15:40:58.529Z]    CovarianceVisual( Ogre::SceneManager* scene_manager, Ogre::SceneNode* parent_node, bool is_local_rotation, bool is_visible = true, float pos_scale = 1.0f, float ori_scale = 0.1f, float ori_offset = 0.1f);

[2019-09-26T15:40:58.529Z]    ^~~~~~~~~~~~~~~~

[2019-09-26T15:40:58.529Z] make[2]: *** [CMakeFiles/fuse_viz.dir/src/mapped_covariance_property.cpp.o] Error 1

[2019-09-26T15:40:58.529Z] make[2]: *** Waiting for unfinished jobs....

[2019-09-26T15:40:58.529Z] In file included from /home/ec2-user/workspace/test_rosdistro_fuse_PR-107/package/fuse_viz/src/relative_pose_2d_stamped_constraint_property.cpp:36:0:

[2019-09-26T15:40:58.529Z] /home/ec2-user/workspace/test_rosdistro_fuse_PR-107/package/fuse_viz/include/fuse_viz/mapped_covariance_visual.h: In constructor ‘rviz::MappedCovarianceVisual::MappedCovarianceVisual(Ogre::SceneManager*, Ogre::SceneNode*, bool, bool, float, float, float)’:

[2019-09-26T15:40:58.529Z] /home/ec2-user/workspace/test_rosdistro_fuse_PR-107/package/fuse_viz/include/fuse_viz/mapped_covariance_visual.h:74:115: error: ‘rviz::CovarianceVisual::CovarianceVisual(Ogre::SceneManager*, Ogre::SceneNode*, bool, bool, float, float, float)’ is private within this context

[2019-09-26T15:40:58.529Z]      : CovarianceVisual(scene_manager, parent_node, is_local_rotation, is_visible, pos_scale, ori_scale, ori_offset)

[2019-09-26T15:40:58.529Z]                                                                                                                    ^

[2019-09-26T15:40:58.529Z] In file included from /home/ec2-user/workspace/test_rosdistro_fuse_PR-107/package/fuse_viz/include/fuse_viz/mapped_covariance_visual.h:33:0,

[2019-09-26T15:40:58.529Z]                  from /home/ec2-user/workspace/test_rosdistro_fuse_PR-107/package/fuse_viz/src/relative_pose_2d_stamped_constraint_property.cpp:36:

[2019-09-26T15:40:58.529Z] /opt/locusrobotics/hotdog/dev/ros1/include/rviz/default_plugin/covariance_visual.h:93:3: note: declared private here

[2019-09-26T15:40:58.529Z]    CovarianceVisual( Ogre::SceneManager* scene_manager, Ogre::SceneNode* parent_node, bool is_local_rotation, bool is_visible = true, float pos_scale = 1.0f, float ori_scale = 0.1f, float ori_offset = 0.1f);

[2019-09-26T15:40:58.529Z]    ^~~~~~~~~~~~~~~~

[2019-09-26T15:40:58.529Z] make[2]: *** [CMakeFiles/fuse_viz.dir/src/relative_pose_2d_stamped_constraint_property.cpp.o] Error 1

[2019-09-26T15:40:58.529Z] In file included from /home/ec2-user/workspace/test_rosdistro_fuse_PR-107/package/fuse_viz/src/relative_pose_2d_stamped_constraint_visual.cpp:36:0:

[2019-09-26T15:40:58.529Z] /home/ec2-user/workspace/test_rosdistro_fuse_PR-107/package/fuse_viz/include/fuse_viz/mapped_covariance_visual.h: In constructor ‘rviz::MappedCovarianceVisual::MappedCovarianceVisual(Ogre::SceneManager*, Ogre::SceneNode*, bool, bool, float, float, float)’:

[2019-09-26T15:40:58.529Z] /home/ec2-user/workspace/test_rosdistro_fuse_PR-107/package/fuse_viz/include/fuse_viz/mapped_covariance_visual.h:74:115: error: ‘rviz::CovarianceVisual::CovarianceVisual(Ogre::SceneManager*, Ogre::SceneNode*, bool, bool, float, float, float)’ is private within this context

[2019-09-26T15:40:58.529Z]      : CovarianceVisual(scene_manager, parent_node, is_local_rotation, is_visible, pos_scale, ori_scale, ori_offset)

[2019-09-26T15:40:58.529Z]                                                                                                                    ^

[2019-09-26T15:40:58.529Z] In file included from /home/ec2-user/workspace/test_rosdistro_fuse_PR-107/package/fuse_viz/include/fuse_viz/mapped_covariance_visual.h:33:0,

[2019-09-26T15:40:58.529Z]                  from /home/ec2-user/workspace/test_rosdistro_fuse_PR-107/package/fuse_viz/src/relative_pose_2d_stamped_constraint_visual.cpp:36:

[2019-09-26T15:40:58.529Z] /opt/locusrobotics/hotdog/dev/ros1/include/rviz/default_plugin/covariance_visual.h:93:3: note: declared private here

[2019-09-26T15:40:58.529Z]    CovarianceVisual( Ogre::SceneManager* scene_manager, Ogre::SceneNode* parent_node, bool is_local_rotation, bool is_visible = true, float pos_scale = 1.0f, float ori_scale = 0.1f, float ori_offset = 0.1f);

[2019-09-26T15:40:58.529Z]    ^~~~~~~~~~~~~~~~

[2019-09-26T15:40:58.529Z] make[2]: *** [CMakeFiles/fuse_viz.dir/src/relative_pose_2d_stamped_constraint_visual.cpp.o] Error 1

[2019-09-26T15:40:58.529Z] make[1]: *** [CMakeFiles/fuse_viz.dir/all] Error 2

[2019-09-26T15:40:58.529Z] make: *** [all] Error 2

[2019-09-26T15:40:58.529Z] ---

[2019-09-26T15:40:58.529Z] Failed   <<< fuse_viz	[ Exited with code 2 ]
```

**Comment by efernandez** [comment]
Thanks, it's that.

Let me know if you prefer me to copy the `CovarianceVisual` code from https://github.com/ros-visualization/rviz/blob/melodic-devel/src/rviz/default_plugin/covariance_visual.h and https://github.com/ros-visualization/rviz/blob/melodic-devel/src/rviz/default_plugin/covariance_visual.cpp, or you'll upgrade `rviz` to 1.13.4.

It's a shame they don't allow to easily compound those classes :disappointed: 
I'd revisit this in a few weeks/months and try to contribute something to `rviz` so we can do this a lot cleaner, but it'd be great if we could get this in now. :)

**Comment by svwilliams** [comment]
I'll take a look at everything in the next couple of days.

**Comment by svwilliams** [comment]
At least for now, I'm trying to maintain compatibility back to Kinetic. So if you need functionality that doesn't go back that far, I'd say you need to duplicate the `CovarianceVisual` code.

Or you can always release fuse_viz as a separate package, outside of the main fuse repo. Then you can add whatever version restrictions you want.

**Comment by efernandez** [comment]
I've reverted the last commit, which is the one that requires `rviz` 1.13.4. Now it should build.
I've also `roslint` the code.

File: fuse_viz/include/fuse_viz/conversions.h [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+{
+  return toOgre(toTF(orientation));
+}
+
+}  // namespace
+
+namespace
+{
+
+inline tf2::Transform getPose(const fuse_core::Graph& graph, const fuse_variables::Position2DStamped& position,
```
Comment (svwilliams) [inline fuse_viz/include/fuse_viz/conversions.h]: I'm not exactly sure the purpose of the `graph` input argument, unless it is for symmetry somewhere.

File: fuse_viz/src/serialized_graph_display.cpp [hunk 2]
```diff
@@ -155,36 +197,111 @@ void SerializedGraphDisplay::processMessage(const fuse_msgs::SerializedGraph::Co
...
-    graph_shapes_.push_back(axes);
+    const auto constraint_uuid = constraint.uuid();
+    const auto& constraint_source = constraint.source();
+
+    if (source_color_map_.find(constraint_source) == source_color_map_.end())
+    {
+      const auto n = source_color_map_.size() + 1;
+      const size_t level = std::floor(std::log2(n));
+      const auto m = n + 1 - std::pow(2, level);
+      const auto hue = (2 * (m - 1) + 1) / std::pow(2, level + 1);
```
Comment (svwilliams) [inline fuse_viz/src/serialized_graph_display.cpp:233]: I feel like there is some black magic here. Mind adding a comment about exactly what you are doing? Creating some sort of pseudo-random color?

**Review by svwilliams — APPROVED** [review]
There is definitely some repetition between the pose visual/property and the relative pose visual/property, but I don't think it's worth refactoring anything at this point.

File: fuse_viz/src/serialized_graph_display.cpp (hunk 2 above)
Comment (efernandez) [reply fuse_viz/src/serialized_graph_display.cpp:233]: I added a comment. I hope I explained it well. I was trying to find some code to generate a dynamic colormap that keeps the colors well spread as the colormap size grows, but I couldn't find anything, so I did some math and came up with this.

File: fuse_viz/include/fuse_viz/conversions.h (hunk 1 above)
Comment (efernandez) [reply fuse_viz/include/fuse_viz/conversions.h]: None. Removed.

**Comment by efernandez** [comment]
@svwilliams I addressed your comments and rebased this on top of the current `devel`.

**Review by matthewbwhitaker — APPROVED** [review]

**Pull request merged**